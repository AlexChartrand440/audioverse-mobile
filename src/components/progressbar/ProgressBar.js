import React from 'react'
import { View, Text, TouchableWithoutFeedback, Slider, Platform, StyleSheet } from 'react-native'
import TrackPlayer, { ProgressComponent } from 'react-native-track-player'

import { formatTime } from 'src/utils'
import flatDot from 'assets/flat_dot.png'

class ProgressBar extends ProgressComponent {

  handleValueChange = (value) => {
    TrackPlayer.pause()
    TrackPlayer.seekTo(value)
  }

  handleSlidingComplete = () => {
    TrackPlayer.play()
  }

  render() {
    const position = formatTime(Math.floor(this.state.position))
    const duration = formatTime(Math.floor(this.state.duration))
    const info = position + ' / ' + duration

    return (
      <View style={styles.container}>
        <TouchableWithoutFeedback>
          <Slider
            value={this.state.position}
            minimumValue={0}
            maximumValue={this.state.duration}
            step={1}
            minimumTrackTintColor="#E53935"
            thumbTintColor="#E53935"
            thumbImage={flatDot}
            style={styles.slider}
            onValueChange={this.handleValueChange}
            onSlidingComplete={this.handleSlidingComplete} />
        </TouchableWithoutFeedback>
        <View style={styles.info}>
          <Text style={{marginLeft: 10}}>{position}</Text>
          <Text style={{marginRight: 10}}>{duration}</Text>
        </View>
      </View>
    )
  }

}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    flexDirection: 'column',
    alignItems: 'center'
  },
  slider: {
    width: '100%',
    marginBottom: Platform.OS === 'ios' ? -20 : -10,
    zIndex: 10
  },
  info: {
    backgroundColor: '#E0E0E080',
    width: '100%',
    paddingTop: Platform.OS === 'ios' ? 10 : 5,
    flexDirection: 'row',
    justifyContent: 'space-between',
  }
})

export default ProgressBar
