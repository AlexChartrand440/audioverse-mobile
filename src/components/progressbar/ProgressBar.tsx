import Slider from '@react-native-community/slider';
import React from 'react';
import { Platform, StyleSheet, Text, TouchableWithoutFeedback, View } from 'react-native';
import TrackPlayer, { useProgress } from 'react-native-track-player';

import flatDot from '../../../assets/flat_dot.png';
import I18n from '../../../locales';
import { primaryColor } from '../../styles';
import { formatTime } from '../../utils';

interface Props {
	rate: number;
}

const styles = StyleSheet.create({
	container: {
		width: '100%',
		flexDirection: 'column',
		alignItems: 'center',
	},
	slider: {
		width: '100%',
		marginBottom: Platform.OS === 'ios' ? -20 : -10,
		zIndex: 10,
	},
	info: {
		backgroundColor: '#E0E0E080',
		width: '100%',
		paddingTop: Platform.OS === 'ios' ? 10 : 5,
		flexDirection: 'row',
		justifyContent: 'space-between',
	},
});

const ProgressBar: React.FC<Props> = ({ rate }) => {
	const progress = useProgress();

	const handleValueChange = async (value: number) => {
		await TrackPlayer.pause();
		await TrackPlayer.seekTo(value);
	};

	const handleSlidingComplete = async () => {
		await TrackPlayer.play();
		// workaround on iOS play/pause resets the playback speed to 1
		// https://github.com/react-native-kit/react-native-track-player/issues/614
		if (Platform.OS === 'ios') {
			if (rate !== 1) {
				await TrackPlayer.setRate(rate);
			}
		}
	};

	const position = formatTime(Math.floor(progress.position));
	const duration = formatTime(Math.floor(progress.duration));

	return (
		<View style={styles.container}>
			<TouchableWithoutFeedback>
				<Slider
					value={progress.position}
					minimumValue={0}
					maximumValue={progress.duration}
					step={1}
					minimumTrackTintColor={primaryColor}
					thumbTintColor={primaryColor}
					thumbImage={Platform.OS === 'ios' ? flatDot : null}
					style={styles.slider}
					onValueChange={handleValueChange}
					onSlidingComplete={handleSlidingComplete}
				/>
			</TouchableWithoutFeedback>
			<View style={styles.info}>
				<Text style={{ marginLeft: 10 }} accessibilityLabel={`${I18n.t('current_time')} ${position}`}>
					{position}
				</Text>
				<Text style={{ marginRight: 10 }} accessibilityLabel={`${I18n.t('duration')} ${duration}`}>
					{duration}
				</Text>
			</View>
		</View>
	);
};

export default ProgressBar;
