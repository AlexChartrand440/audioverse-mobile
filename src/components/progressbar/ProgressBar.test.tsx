import React from 'react'
import ProgressBar from './ProgressBar'

import renderer from 'react-test-renderer'

test('renders correctly', () => {
  const tree = renderer.create(<ProgressBar rate={1} />).toJSON()
  expect(tree).toMatchSnapshot()
})
