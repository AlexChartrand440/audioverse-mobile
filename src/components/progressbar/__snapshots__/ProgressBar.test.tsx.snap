// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <RCTSlider
    accessible={true}
    disabled={false}
    enabled={true}
    maximumValue={1}
    minimumTrackTintColor="#E53935"
    minimumValue={0}
    onChange={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={[Function]}
    step={1}
    style={
      Array [
        Object {
          "height": 40,
        },
        Object {
          "marginBottom": -20,
          "width": "100%",
          "zIndex": 10,
        },
      ]
    }
    thumbImage={1}
    thumbTintColor="#E53935"
    value={0}
  />
  <View
    style={
      Object {
        "backgroundColor": "#E0E0E080",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <Text
      accessibilityLabel="current_time "
      style={
        Object {
          "marginLeft": 10,
        }
      }
    >
      
    </Text>
    <Text
      accessibilityLabel="duration "
      style={
        Object {
          "marginRight": 10,
        }
      }
    >
      
    </Text>
  </View>
</View>
`;
